File: fe/src/types/seminar.ts
export interface SeminarResponse {
  id: number;
  created_at: string;
  title: string;
  description: string | null;
  start_time: string | null;
  end_time: string | null;
  location: string | null;
  max_capacity: number | null;
  host: string | null;
  cover_image: string | null;
  rsvp_enabled: boolean;
  waitlist_enabled: boolean;
  /** IANA timezone the organiser selected when creating the seminar, e.g. "America/Los_Angeles". */
  display_timezone: string | null;
}

export interface SeminarUserResponse {
  id: number;
  email: string;
  username: string;
  checked_in: boolean;
  checked_in_at: string | null;
}

export interface SeminarWaitlistUserResponse {
  id: number;
  email: string;
  username: string;
  position: number;
  joined_at: string;
}

export interface SeminarDetailResponse extends SeminarResponse {
  display_timezone: string | null;
  current_rsvp_count: number;
  waitlist_count: number;
  users: SeminarUserResponse[];
  waitlist: SeminarWaitlistUserResponse[];
}

export interface CheckInTokenResponse {
  id: number;
  seminar_id: number;
  token: string;
  expires_at: string;
  is_active: boolean;
}

export interface RsvpResponse {
  message: string;
  waitlisted: boolean;
  position?: number;
}

File: be/src/model/seminar.py
from sqlalchemy.orm import Mapped, mapped_column, relationship
from src.config.database import Base
from sqlalchemy import func, DateTime, Boolean, String, Integer
from datetime import datetime
from typing import Optional


class Seminar(Base):
    __tablename__ = "seminars"
    id: Mapped[int] = mapped_column(primary_key=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    title: Mapped[str] = mapped_column()
    description: Mapped[Optional[str]] = mapped_column(nullable=True)
    start_time: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    end_time: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    location: Mapped[Optional[str]] = mapped_column(nullable=True)
    max_capacity: Mapped[Optional[int]] = mapped_column(nullable=True)
    host: Mapped[Optional[str]] = mapped_column(nullable=True)
    cover_image: Mapped[Optional[str]] = mapped_column(nullable=True)
    rsvp_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    waitlist_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    # IANA timezone used when creating the event (e.g. "America/Los_Angeles").
    # Stored so the frontend can display times back in the organiser's intended timezone.
    display_timezone: Mapped[Optional[str]] = mapped_column(String, nullable=True)

    rsvps = relationship(
        "SeminarRSVP",
        backref="seminar",
        cascade="all, delete-orphan"
    )
    waitlist = relationship(
        "SeminarWaitlist",
        backref="seminar",
        cascade="all, delete-orphan",
        order_by="SeminarWaitlist.created_at"
    )
    checkin_tokens = relationship(
        "CheckInToken",
        backref="seminar",
        cascade="all, delete-orphan"
    )
